signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,446 @@
+
+        if self.pending_connections.remove(&connection_id) {
+            tracing::debug!(%connection_id, %peer_id, %address, "successfully dialed seed peer");
+            self.pending_events
+                .push_back(ToSwarm::GenerateEvent(BootstrapEvent::SeedConnected {
+                    connection_id,
+                    peer_id,
+                    address: address.clone(),
+                }));
+        }
+
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn handle_pending_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        maybe_peer: Option<PeerId>,
+        addresses: &[Multiaddr],
+        _effective_role: Endpoint,
+    ) -> Result<Vec<Multiaddr>, ConnectionDenied> {
+        let addresses = addresses.to_vec();
+
+        // If the connection ID is in the pending connections set, then we know
+        // that this is a seed peer dial that we've initiated, so we publish
+        // the `DialingSeed` event.
+        if self.pending_connections.contains(&connection_id) {
+            tracing::debug!(%connection_id, addresses = ?addresses, peer_id = ?maybe_peer, "attempting to dial seed peer");
+            self.pending_events
+                .push_back(ToSwarm::GenerateEvent(BootstrapEvent::DialingSeed {
+                    connection_id,
+                    addresses: addresses.clone(),
+                    peer_id: None,
+                }));
+        }
+
+        // Return the addresses as-is.
+        Ok(addresses)
+    }
+
+    fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm) {
+        match event {
+            FromSwarm::ConnectionClosed(e) => {
+                let (peer_id, connection_id, address) =
+                    (e.peer_id, e.connection_id, e.endpoint.get_remote_address());
+                self.peer_disconnected(peer_id, connection_id, address);
+            }
+            FromSwarm::DialFailure(e) => {
+                let (connection_id, error, peer_id) = (e.connection_id, e.error, e.peer_id);
+                if self.pending_connections.remove(&connection_id) {
+                    tracing::debug!(
+                        %connection_id,
+                        ?peer_id,
+                        %error,
+                        "failed to dial seed peer"
+                    );
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: PeerId,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _event: libp2p::swarm::THandlerOutEvent<Self>,
+    ) {
+    }
+
+    #[tracing::instrument(skip_all)]
+    fn poll(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p::swarm::ToSwarm<Self::ToSwarm, libp2p::swarm::THandlerInEvent<Self>>>
+    {
+        // If we have any pending events, we return them immediately.
+        if let Some(event) = self.pending_events.pop_front() {
```

Comment:
nit: use `self.next_pending_event()`? It may end up _uglier_, but at least we are using the same interface

Reply:
Sure: [f982da0](https://github.com/stacks-network/sbtc/pull/1257/commits/f982da0952823fa72b602afd620ff364ca473374)